Client/Frontend/Browser/Playlist/Cache/PlaylistDownloadManager.swift

```diff
@@ -86,20 +85,67 @@ public class PlaylistDownloadManager: PlaylistStreamDownloadManagerDelegate {
     }
     
     func downloadHLSAsset(_ assetUrl: URL, for item: PlaylistInfo) {
-        hlsDelegate.downloadAsset(hlsSession, assetUrl: assetUrl, for: item)
+        if Thread.current.isMainThread {
+            hlsDelegate.downloadAsset(self.hlsSession, assetUrl: assetUrl, for: item)
+        } else {
+            hlsQueue.addOperation {  [weak self] in
+                guard let self = self else { return }
+                self.hlsDelegate.downloadAsset(self.hlsSession, assetUrl: assetUrl, for: item)
+            }
+        }
     }
     
     func downloadFileAsset(_ assetUrl: URL, for item: PlaylistInfo) {
-        fileDelegate.downloadAsset(fileSession, assetUrl: assetUrl, for: item)
+        if Thread.current.isMainThread {
+            fileDelegate.downloadAsset(self.fileSession, assetUrl: assetUrl, for: item)
+        } else {
+            fileQueue.addOperation {  [weak self] in
+                guard let self = self else { return }
+                self.fileDelegate.downloadAsset(self.fileSession, assetUrl: assetUrl, for: item)
+            }
+        }
     }
     
     func cancelDownload(item: PlaylistInfo) {
-        hlsDelegate.cancelDownload(item: item)
-        fileDelegate.cancelDownload(item: item)
+        if Thread.current.isMainThread {
+            hlsDelegate.cancelDownload(item: item)
+            fileDelegate.cancelDownload(item: item)
+        } else {
+            hlsQueue.addOperation { [weak self] in
```

Comment:
For canceling the operation cant we just force it to run on main thread directly ?

Reply:
It is ran on main-thread only atm. This is only if there is a different thread (IE: I left it there for when debugging performance issues on separate threads).